.gitmodules

```diff
@@ -0,0 +1,3 @@
+[submodule "external/cpu_features"]
```

Comment:
I am not a super big fan of submodules. Cleaning things up is good but can we go for `git subtree` instead? That allows us to still track upstream but keep all source in tree. When importing the external source via `git subtree` we should squash the history to avoid it messing up ours. Using `git subtree` allows us to do changes on top which gets hard with submodules.

Reply:
Interesting! I did not know about [`git-subtree`](https://github.com/git/git/blob/master/contrib/subtree/git-subtree.txt) before.

To me `git-subtree` seems to require more attention, as one needs to manually make sure not to change the subtree and other files in the same commit. Furthermore, we would lose the commit history of the project by sqashing.

I think submodules make it easier to adapt upstream changes, as we just need to reference another upstream commit. Should we need to incorporate changes which will not be accepted upstream, we could switch to a fork, keeping maintainance of those patches seperate and easier to adapt to new upstream versions.

Keeping the projects seperated via submodules would lower the maintenance burden in my opinion, while `git-subtrees` mixes the projects together too much.

I might have not understood your approach correctly though, as I have not used `git-subtrees` before, so feel free to correct me.
In the end it is your decision though and I would change the PR if you wish so.